ably/rest_channel.go

```diff
@@ -53,6 +53,14 @@ func newRESTChannel(name string, client *REST) *RESTChannel {
 
 // Publish publishes a message on the channel.
 func (c *RESTChannel) Publish(ctx context.Context, name string, data interface{}) error {
+	//Check to see if data received is of type message.
+	if msg, ok := data.(Message); ok {
```

Comment:
Actually, I am skeptical about `Message` type being used as a public struct. My understanding for the same is, it's supposed to hold message data. Other metadata is modified internally, like `timestamp`, `connectionId`, `clientID` etc.
So, we shouldn't be passing it from outside. Also `data interface {}` as a struct can hold any type of data, adding `connectionKey` to the struct object shouldn't force to use `Message` type. Also, it feels redundant to use `connectionKey` as a part of an actual data message. My personal opinion would be to add extra param to Publish. Since, go doesn't support optional params, as per suggestions by go devs, a new method can be added like this
`PublishWithConnectionKey(ctx context.Context, name string, data interface{}, connectionKey string)` or 
`PublishOnBehalfOf(ctx context.Context, name string, data interface{}, connectionKey string)`. The second name seems more verbose and easy to understand in the public interface itself, instead of making things confusing by including `connectionKey` inside `data` ( user needs to look at documentation explicitly). 

The only issue I see with this is, it's not defined as a public interface in the spec. Though, go language limits reuse of same interface method due to lack of optional param support.

Reply:
@lmars I would like your input for the same, also you can point out if I am missing something.

Reply:
I have reworked the implementation of this bug fix to use a functional option to provide the connection key.

Reply:
@sacOO7 please can you resolve this conversation if you are happy that this code has been re-worked to use functional options .

Reply:
Resolving this conversation on @sacOO7's behalf so that @Rosalita can land this PR.